### PR TITLE
Add Perfomance/Detect cop

### PR DESCRIPTION
**Pull request by palkan**

### CODE REVIEW

File: lib/rubocop/cop/performance/detect.rb [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+module RuboCop
+  module Cop
+    module Performance
+      # This cop is used to identify usages of `select.first` and
+      # change them to use `detect` instead.
+      #
+      # @example
+      #   # bad
+      #   [].select { |item| true }.first
```
Comment (bbatsov) [inline lib/rubocop/cop/performance/detect.rb]: This should handle `find_all` as well (it's an alias of select).

**Comment by bbatsov** [comment]
One thing to consider - some people prefer `find` over `detect`, so there should be some configuration that would affect the message and the autocorrect.

**Comment by bbatsov** [comment]
:+1: Looks good. Great work!

File: config/default.yml [hunk 2]
```diff
@@ -711,3 +711,10 @@ Rails/TimeZone:
 Rails/Validation:
   Include:
     - app/models/**/*.rb
+
+##################### Performance ##################################
+
+Performance/Detect:
+  # Defines what method to use instead of `select.first`, e.g.
+  # `detect` or `find`.
+  PreferredMethod: detect
```
Comment (jonas054) [inline config/default.yml]: Isn't it better to use the configuration parameter `Style/CollectionMethods:PreferredMethods` to choose between  `detect` and `find`?

File: lib/rubocop/cop/performance/detect.rb [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+          receiver, _args, _body = *receiver if receiver.block_type?
+
+          _, first_method = *receiver
+          return unless SELECT_METHODS.include?(first_method)
+
+          source_buffer = node.loc.expression.source_buffer
+          location_of_select = receiver.loc.selector.begin_pos
+          end_location = node.loc.selector.end_pos
+
+          range = Parser::Source::Range.new(source_buffer,
```
Comment (jonas054) [inline lib/rubocop/cop/performance/detect.rb]: `Range#join` is an easier way to get from the beginning of one range to the end of the next. So, `receiver.loc.selector.join(node.loc.selector)`.

File: lib/rubocop/cop/performance/detect.rb [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+          add_offense(node, range, format(MSG,
+                                          preferred_method,
+                                          first_method))
+        end
+
+        def autocorrect(node)
+          receiver, _first_method = *node
+
+          first_range = Parser::Source::Range.new(
```
Comment (jonas054) [inline lib/rubocop/cop/performance/detect.rb]: Use `Range#join`.

File: lib/rubocop/cop/performance/detect.rb [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+            node.loc.expression.source_buffer,
+            node.loc.dot.begin_pos,
+            node.loc.selector.end_pos)
+
+          receiver, _args, _body = *receiver if receiver.block_type?
+
+          select_range = Parser::Source::Range.new(
```
Comment (jonas054) [inline lib/rubocop/cop/performance/detect.rb]: This is equivalent to `select_range = receiver.loc.selector`.

File: lib/rubocop/cop/performance/detect.rb [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+          @corrections << lambda do |corrector|
+            corrector.remove(first_range)
+            corrector.replace(select_range, preferred_method)
+          end
+        end
+
+        private
+
+        def preferred_method
+          cop_config['PreferredMethod'] || default_cop_config['PreferredMethod']
```
Comment (jonas054) [inline lib/rubocop/cop/performance/detect.rb]: The default configuration is already "baked in" to `cop_config`.

**Comment by jonas054** [comment]
@palkan Some late comments for you. Can you take a look?

File: config/default.yml (hunk 2 above)
Comment (palkan) [inline config/default.yml]: I think it's against single responsibility, isn't it?

**Comment by palkan** [comment]
@jonas054 
Opened new PR with fixes - #1715.

File: config/default.yml (hunk 2 above)
Comment (jonas054) [inline config/default.yml]: No, not at all. It is impossible to prefer `find` over `detect` in general but still want `select.first` to be changed to `detect`. So it's a single concern, and specifying it in two places is a violation of the DRY principle.

Using another cop's configuration is something we do in a few places already. Search for `for_cop` in the source code.

**Comment by basex** [comment]
I had a problem with this cop with auto-correction of Datamapper select, from adapter.select() to adapter.find(), that doesn't exist in DM. 

Can we only do the auto-correct if the word "adapter" doesn't exist before the select?

**Comment by palkan** [comment]
Please, provide full code snippet.

**Comment by basex** [comment]
This was the diff:
-   DataMapper.repository(:default).adapter.select('select id from ....')
+    DataMapper.repository(:default).adapter.find('select id from ....')

**Comment by palkan** [comment]
Ok, got it.
Please open a new issue, I'm working on it.

**Comment by basex** [comment]
Thanks a lot. Issue opened #1766

**Comment by bbatsov** [comment]
This can also be a problem with ActiveRecord (as it also has a `select` method). Probably the best approach would be to simply exclude the cop for such files. I don't think this can be handled in the code in any reasonable way (I certainly don't want to support a ton of special cases).

**Comment by palkan** [comment]
@bbatsov 
I think we can) 
`Enumerable#select` requires block (either normal block or ampersand-style block), so we can use this to distinguish from AR/DataMapper and others `select`. See [PR](https://github.com/bbatsov/rubocop/pull/1767).

**Pull request merged**